p4-16/spec/P4-16-spec.mdk

```diff
@@ -2795,6 +2795,37 @@ important what the type variable binds to (during type unification the
 don't care type can unify with any other type).  An example is given
 Section [#sec-arch-desc-example].
 
+## Default values
+
+Some P4 types define a "default value," which can be used to
+automatically initialize values of that type.  The default values are
+as follows:
+
+* For `int`, `bit<N>` and `int<N>` types the default value is 0.
+* For `bool` the default value is `false`.
+* For `error` the default value is `error.NoError` (defined in core.p4)
+* For `string` the default value is the empty string `""`
+* For `varbit<N>` the default value is a string of zero bits (there is currently
+   no P4 literal to represent such a value).
+* For `enum` values with an underlying type the default value is 0,
+  even if 0 is actually not one of the named values in the enum.
+* For `enum` values without an underlying type the default value is the
+  first value that appears in the `enum` type declaration.
+* For `header` types the default value is `invalid`.
```

Comment:
As far as I can see it is impossible to set header validity bit to zero with existing methods of initialization (list expression or struct initializer expression).  Do you find it convenient that you can do it with shorten notation of existing initializations(`= ...`), and with regular ones you cannot?

Reply:
It is my understanding (corrections welcome if I am wrong here) that nothing in this proposal enables another way to assign an invalid header value to a header.  Whether this proposal goes forward as it is now, or not, these are the only ways I know of for a header to become invalid:

+ You make a call to the method `myhdr.setInvalid()`
+ Assign to a header, the value of another variable with the same header type, and that variable is currently an invalid header.
+ direction `out` parameters with a header type are defined to be initialized to invalid, without any explicit statement in the program that does so. (I am not sure whether one can define an `out` parameter with a header stack type -- not something I have tried before or seen brought up).
+ header stack push_front and pop_front operations can change which elements of the header stack are valid or invalid.

Reply:
@marko-stanoj Are you asking "Would others like to introduce another way to make a header invalid, other than the ones that exist now, which uses an initialization expression?"

Reply:
Later in this document it's said:

> A left-value can be initialized automatically with a default value of
> the suitable type using the syntax `...`

so if left-value is header type (or struct type that contains header type as a field) it should be initialized to it's default value. Due to this spec proposal, the default value is invalid, or am I missing something?

Reply:
I hereby withdraw this statement in an earlier comment: "It is my understanding (corrections welcome if I am wrong here) that nothing in this proposal enables another way to assign an invalid header value to a header."

Now upon reading this proposed change, it does appear to me that the syntax `h1 = ...;` is proposed to make a header variable `h1` become invalid, no matter what its state was before, and syntax `h1 = {f1=5, ...}` is to make `h1` valid, with field `f1` becoming 5, and all other fields being assigned their default values, depending upon the type of those fields.

In other comments on this PR, I propose adding explicit examples with comments explaining this, rather than leaving it open to the imagination of the reader.